### ISSUE TITLE
[RFC] Dynamic Hybrid Search Optimization

### ISSUE THREAD

**Issue by wrigleyDan** [opensearch-project/search-relevance#206]
## Introduction

Hybrid search as a technique to combine results from multiple queries is seeing increasing adoption. However, there are numerous tuning knobs users can turn to optimize search quality. 

The goal of this RFC is to describe a path forward for an approach that allows dynamically predicting the best hybrid search parameter combination on a per-query basis with machine learning.

## Background & Motivation

The Search Relevance Workbench introduced a straightforward approach for identifying suitable hybrid search parameters by trying out a hard-coded set of different alternatives with a predefined query set and hybrid search configuration.  
However, this approach produces exactly one parameter combination as “the best” one for all queries. With this parameter combination, some queries will benefit (as in their search quality metrics improve) and other queries will not benefit (as in their search quality metrics decrease) as described in the [“Hybrid Search Optimization” blog post](https://opensearch.org/blog/hybrid-search-optimization/).  
Dynamic hybrid search optimization is a search relevance tuning operation for advanced users that requires machine learning knowledge, specifically around feature selection & engineering processes and model training.

## Proposed Solution

In a nutshell, predicting the best parameters on a per-query level with a machine learning model has the requirements that all ML-powered applications have:

* Feature selection & feature engineering  
* Model training & serving  
* Model inference

This will be an advanced feature with an expected high level of individuality per user:

* Not all features will be useful for all user applications.  
* Feature distributions will be different across different user applications.  
* Different user applications have different non-functional requirements that have an impact on feature selection.  
* Machine learning algorithms have a set of hyperparameters to tune and the ideal parameter values will be dependent on factors that are individual per user application.

A strong experimentation mindset helps identify the best features & parameters.

### Feature selection & engineering

We anticipate users wanting to train their own models including feature selection & feature engineering. While latency-sensitive search applications like e-commerce search may want to use fewer features related to the user query only, other applications like RAG systems may include search result features that require executing queries and using resulting information as features.

### Model training and serving

Users should be able to use suitable models supported by OpenSearch, mainly [linear regression](https://docs.opensearch.org/docs/latest/ml-commons-plugin/algorithms/#linear-regression), and also host models outside of the OpenSearch context which can be called from within queries.

### Model inference

We assume that model inference needs to be flexible:

* Users may want to train models that predict the best weight for arithmetic neural/keyword search combinations.  
* Users may want to train models that predict the best ranking constant to be used in an RRF setting.  
* Users may want to train models that predict the best combination technique, normalization technique and the corresponding weights for the queries.

The response format needs to somehow be mapped to a corresponding parameter (value) in a Query DSL object.

### Feature selection & engineering

We expect feature selection and engineering to be a process outside of OpenSearch handled with tools and frameworks that excel in this space like Scikit-learn, TensorFlow, PyTorch or Keras.

### Model training and serving

[OpenSearch provides a built-in machine learning algorithm for linear regression.](https://docs.opensearch.org/docs/latest/ml-commons-plugin/algorithms/#linear-regression)   
For other algorithms or more complex services users can train models external to OpenSearch and run them in custom services, for example hosted on Amazon SageMaker that can be connected to OpenSearch platforms via [external connectors](https://docs.opensearch.org/docs/latest/ml-commons-plugin/remote-models/index/).

### Model inference

The `ml_inference` search pipeline can be used to connect to internally or externally hosted models. It is known that externally hosted models can significantly add to latency which may be too high for OpenSearch users. An initial experimentation phase will show the detailed impact and reveal the need for other solutions like internally hosted models similar to LTR - coming with its own limitations (not being able to run on the serverless platform).  
Feature values are passed into the model and predicted values are retrieved from the `ml_inference` call. Queries can already be defined in `ml_inference` search pipelines via the `query_template` parameter.  
However, one current limitation is that it is not possible to use the output of the `ml_inference` pipeline in the defined query (see [https://github.com/opensearch-project/neural-search/issues/1172](https://github.com/opensearch-project/neural-search/issues/1172)).

In short, the output values of the `ml_inference` pipeline step need to be passed to the corresponding parts of the hybrid search query, specifically the `phase_results_processors`.

Example search pipeline:  
```
PUT _search/pipeline/ml_inference_pipeline  
{  
  "description": "search with predictions",  
  "request_processors": [  
    {  
      "ml_inference": {  
        "function_name": "remote",  
        "full_response_path": true,  
        "model_id": "zRG2YpMBDpIvIC8uJxhB",  
        "model_input": """{ "parameters": {"input": "${input_map.features}"}}""",  
        "query_template": """{  
          "_source": {  
            "excludes": [  
              "title_embedding"  
            ]  
          },  
          "query": {  
            "hybrid": {  
              "queries": [  
                {  
                  "multi_match": {  
                    "type": "best_fields",  
                    "fields": [  
                      "my_field_a^20",  
                      "my_field_b^10"  
                    ],  
                    "operator": "and",  
                    "query": "iphone"  
                  }  
                },  
                {  
                  "neural": {  
                    "title_embedding": {  
                      "query_text": "iphone",  
                      "k": 50  
                    }  
                  }  
                }  
              ]  
            }  
          },  
         "search_pipeline": {  
           "phase_results_processors": [  
             {  
              "normalization-processor": {  
                "normalization": {  
                "technique": "l2"  
              },  
              "combination": {  
              "technique": "arithmetic_mean",  
              "parameters": {  
                "weights": [  
                  ${keywordness},  
                  ${neuralness}  
                ]  
              }  
            }  
          }  
        }  
      ]  
    },  
    "size": 100  
    }""",  
        "input_map": [  
          {  
            "features": "query.term.features.value"  
          }  
        ],  
        "output_map": [  
          {  
            "neuralness": "$.inference_results[0]output[0]dataAsMap.neuralness",  
            "keywordness": "$.inference_results[0]output[0]dataAsMap.keywordness"  
          }  
        ],  
        "ignore_missing": false,  
        "ignore_failure": false  
      }  
    },  
    {  
      "neural_query_enricher": {  
        "description": "one of many search pipelines for experimentation",  
        "default_model_id": "i6jHTZMBflg_ePyfu9EK",  
        "neural_field_default_id": {  
          "title_embeddings": "i6jHTZMBflg_ePyfu9EK"  
        }  
      }  
    }  
  ]  
}  
```

In this example, the query template holds the hybrid search query and references the parameters that need to be substituted by the predicted values from the ML model (`${keywordness}` & `${neuralness}`).

In the `ml_inference` pipeline processor a remote model is configured. It takes the feature values as its input (value of the parameter `model_input`) and returns two values that are mapped to the previously mentioned parameters in `output_map`.

In this example the user provides the calculated input features. Ideally, there is support for simple features that rely on textual or numerical input, for example the length of a query string or whether a query contains numbers or other specific characters. A first implementation takes features values as an input and a potential next step might be supporting functions in a scripting language like Python.

### When to use dynamic hybrid search optimization

As outlined previously, hybrid search optimization is an advanced search relevance tuning feature for experienced users. Experimental data from the Search Relevance Workbench’s (static) hybrid search optimizer is seen as a good way to identify the potential the dynamic approach may give:  
Experimental results that show a significant amount of queries that profit from different hybrid search settings can point towards promising results for the dynamic hybrid search optimization approach as this means there is no clear “one size fits all” solution.  
In general, this approach will most likely show good results the more diverse and heterogeneous the query distribution is.

Collaborators on this RFC:
* [Martin Gaievski](https://github.com/martin-gaievski)

Related issues & work:

* [https://opensearch.org/blog/hybrid-search-optimization/](https://opensearch.org/blog/hybrid-search-optimization/)   * 
* [Testing & Prototyping in OpenSearch with different datasets](https://github.com/martin-gaievski/info-retrieval-test/tree/dynamic_optimizer_feature_eng_esci_dataset/dynamic_hybrid)
* [https://github.com/opensearch-project/neural-search/issues/1172](https://github.com/opensearch-project/neural-search/issues/1172)   
* [https://github.com/opensearch-project/ml-commons/issues/2841](https://github.com/opensearch-project/ml-commons/issues/2841)  
* [https://github.com/opensearch-project/ml-commons/issues/3210](https://github.com/opensearch-project/ml-commons/issues/3210)  
* [https://github.com/opensearch-project/neural-search/issues/1005](https://github.com/opensearch-project/neural-search/issues/1005)   
* [https://github.com/opensearch-project/OpenSearch/issues/17432](https://github.com/opensearch-project/OpenSearch/issues/17432)   
* [https://github.com/opensearch-project/OpenSearch/pull/18095](https://github.com/opensearch-project/OpenSearch/pull/18095)

**Comment by onestardao** [comment]
your approach here sounds solid on the surface — especially the structured template and model-driven weight tuning — but just a heads-up, we’ve seen a recurring problem in dynamic pipelines like this: when ML feature prediction and search weights aren't semantically aligned, the hybrid query ends up degenerating relevance rather than boosting it.

in our research we catalog this as a typical "weight misalignment under semantic shift" case (ProblemMap No.13).
if you're curious, we've open-sourced a full diagnostic toolset under MIT license — and even got tesseract.js creator to star the repo.
we cover 16 common failure cases in hybrid / RAG / semantic retrieval systems.

if you’re interested, happy to share the link.

**Comment by martin-gaievski** [comment]
@wrigleyDan if I want to start from a simpler version of optimizer that has limited set of most effective query features I see few challenges. In [Optimizing hybrid search in OpenSearch]( https://opensearch.org/blog/hybrid-search-optimization/ ) blog post you've mentioned three feature groups: base query, lexical search and neural search. For lexical and neural search features user needs to know characteristics of dataset, like fields, their types, importance (e.g. for lexical we build features around field "title"). How this can this be generalized for any dataset? I can think of few approaches each with different tradeoffs: 
1. analyze field types (index mapping) and build query dynamically, using some sort of heuristic (something simple like looking for fields of text type and name "title"), then assign different boost factors based on how we categorize fields.
2. Run test query against all fields using OpenSearch query syntax "*". This is way simpler, but all fields will have equal weight, so can be less effective.
 
Another question is regarding judgment ratings, are we going to solely rely on user provided ratings? I know even for ESCI dataset we do have ~30% of coverage, which is not ideal, other datasets are going worse. In theory we need ratings for every pair of query text and document, this can be a lot of efforts to get this. One option would be to generate ratings with LLM, SRW has this functionality. But that probably will be of questionable quality and can be expensive to end user (high number of call to LMM and system resource utilization). Are you aware of any technique that we can use to minimize this, maybe sampling, but I hardly imagine that doesn't affect quality of model training.

**Comment by wrigleyDan** [comment]
> One option is - in initial version we can rely on user provided judgments, similar to query features. Users can generate LLM judgments using current SRW functionality, but we're not going to make it part of the model training workflow.

Exactly, I would not integrate that as part of the workflow but rather let users create their judgment lists in SRW or upload them when created outside and let the dynamic hybrid search optimizer use the available lists instead of trying to generate these on the fly. That indeed would be time- and resource-consuming.

I'm not sure I'm entirely understand your concerns about needing all documents of the entire corpus judged per query. Can you elaborate on that? Usually, having something around the top 20 up to the top 50 per query for a given search configuration already provides you with a decent amount when measuring metrics at result list depth 10, for example.

**Comment by martin-gaievski** [comment]
> I'm not sure I'm entirely understand your concerns about needing all documents of the entire corpus judged per query. Can you elaborate on that? Usually, having something around the top 20 up to the top 50 per query for a given search configuration already provides you with a decent amount when measuring metrics at result list depth 10, for example.

My concern is how to get that top 50 per query. Aren't we suppose to judge on every document in the dataset and then take top 50? If yes this means for A documents and B queries we need to get AxB judgments.

**Comment by martin-gaievski** [comment]
I published RFC for initial version implementation, it's focused more on low level/technical aspects of implementation. Feel free to review and comment
